enhancements/generic-constraints.md

```diff
@@ -0,0 +1,181 @@
+
+### User Stories
+
+#### Generic Constraint
+
+As an operator author, I would like to specify a constraint that has my own rules/requirements that are appliable to a set of known properties are from other bundles or from the cluster itself without having to tell OLM how to evaluate those rules. For example, if I know there is a property named certified in some bundles, I can express a constraint with a rule that says my operator to only be installed along with those certified bundles without having to tell OLM to support a new type of constraint for certified property.
+
+#### Compound Constraint
+
+As an operator author, I would like to specify a constraint that has multiple rules and requirements that must be resolved into a single dependent operator. For example, I would like to have a constraint for a dependent operator that belongs to a specific package and provides a specific API/GVK.
+
+### Design Details
+
+#### Generic Constraint Syntax
+
+The constraint syntax is designed to have similarity with the existing constraints that OLM supports which are GVK and package. It is intentional to keep the fundamental syntax with `type` and `value` the same so that it doesn't introduce any dramatic changes to the overall user experiences.
+
+Each constraint is specified using the following syntax:
+
+```json=
+{
+   "type":"olm.constraint",
+   "value":{
+      "evaluator":{
+         "id":"cel"
+      },
+      "rule": "properties.exists(p, p.type == \"certified\")",
+      "message": "require to have certified property",
+      "action":{
+         "id":"require"
+      }
+   }
+}
+```
+
+The constraint `type` is `olm.constraint` and the `value` is information associated with constraint. The `value` struct has 4 fields: `evaluator`, `rule`, `message` and `action`.
+
+The `evaluator` is a struct with `id` field to represent the language library that will be used to evaluate the expression. At the moment, only CEL expression is supported using `cel` as the identifier.
+
+The `rule` field is the string that presents a CEL expression that will be evaluated during resolution against . Only CEL expression is supported in the initial release.
+
+The `message` field is an optional field that is accommodating the rule field to surface information regarding what this CEL rule is about. The message will be surfaced in resolution output when the rule evaluates to false.
+
+Finally, the `action` field is a struct with the `id` field to indicate the resolution action that this constraint will behave. For example, for `require` action, there must be at least one candidate that satisfies this constraint. This action can be used to indicate optional constraint in the future. For the initial release, `require` action is supported.
+
```

Comment:
General questions:
1. Which of these fields are required vs optional?
2. Are we expecting that _all_ future constraint types can be expressed by `olm.constraint`?
3. If I later want to add another native constraint (e.g. propertyExists), how would that look?

If the answers to 2) and 3) are "no, we'll make a separate root `type` (e.g. `olm.constraint.propertyExists`)", then what is the stance on version skew when a bundle contains a property (that's actually intended to be a constraint) that is unknown to OLM?